src/data_structures.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Common data structures used in this charm."""
+
+import typing
+
+import pydantic
+from charms.bind.v0.dns_record import RecordClass, RecordType, RequirerEntry
+
+
+class DnsEntry(pydantic.BaseModel):
+    """Class used to represent a DNS entry.
+
+    Attributes:
+        domain: example: "dns.test"
+        host_label: example: "admin"
+        ttl: example: 600
+        record_class: example: "IN"
+        record_type: example: "A"
+        record_data: example: "42.42.42.42"
+    """
+
+    domain: str = pydantic.Field(min_length=1)
+    host_label: str = pydantic.Field(min_length=1)
+    ttl: int
+    record_class: RecordClass
+    record_type: RecordType
+    record_data: pydantic.IPvAnyAddress
+
+    def conflicts(self, other: "DnsEntry") -> bool:
+        """Check if it conflicts with another DnsEntry.
+
+        Args:
+            other: DnsEntry to check conflicts with.
+
+        Returns:
+            True if they conflict, False otherwise.
+        """
+        key_attributes: typing.Set[str] = {"domain", "host_label", "record_class", "record_type"}
```

Comment:
Nitpick:
I've received a few feedback saying that I could just do `set` rather than import from `typing` module :) Would this need to support older versions of Python than 3.9 (see deprecation note of https://docs.python.org/3/library/typing.html#typing.Set)

Reply:
If that's true, then we don't need typing of list/dict/etc... ?

Reply:
I've decided that I would make this change in another PR, but thanks for noticing it !